crates/ruff_linter/src/rules/ruff/rules/parenthesize_logical_operators.rs

```diff
@@ -0,0 +1,81 @@
+use ruff_diagnostics::{Diagnostic, Violation};
+use ruff_macros::{derive_message_formats, violation};
+use ruff_python_ast as ast;
+use ruff_python_ast::parenthesize::parenthesized_range;
+use ruff_text_size::Ranged;
+
+use crate::checkers::ast::Checker;
+
+/// ## What it does
+/// Checks for chained operators where adding parentheses could improve the
+/// clarity of the code.
+///
+/// ## Why is this bad?
+/// `and` always binds more tightly than `or` when chaining the two together,
+/// but this can be hard to remember (and sometimes surprising).
+/// Adding parentheses in these situations can greatly improve code readability,
+/// with no change to semantics or performance.
+///
+/// For example:
+/// ```python
+/// a, b, c = 1, 0, 2
+/// x = a or b and c
+///
+/// d, e, f = 0, 1, 2
+/// y = d and e or f
+/// ```
+/// Use instead:
+/// ```python
+/// a, b, c = 1, 0, 2
+/// x = a or (b and c)
+///
+/// d, e, f = 0, 1, 2
+/// y = (d and e) or f
+
+#[violation]
+pub struct ParenthesizeChainedOperators;
+
+impl Violation for ParenthesizeChainedOperators {
+    #[derive_message_formats]
+    fn message(&self) -> String {
+        format!(
+            "Parenthesize `foo and bar` expressions when chaining `and` and `or` together, to make the precedence clear"
+        )
+    }
+}
+
+pub(crate) fn parenthesize_chained_logical_operators(
+    checker: &mut Checker,
+    expr: &ast::ExprBoolOp,
+) {
+    for condition in &expr.values {
+        match condition {
+            // If we find a BoolOp expression inside a BoolOp expression,
+            // it means a different operator is being used for the subexpression
+            // than in the superexpression:
+            // `a or b or c` => `BoolOp(values=[Name('a'), Name('b'), Name('c')], op=Or)`
+            // `a or b and c` => `BoolOp(values=[Name('a'), BoolOp(values=[Name('b'), Name('c')], op=And)], op=Or)`
+            ast::Expr::BoolOp(
+                bool_op @ ast::ExprBoolOp {
+                    op: ast::BoolOp::And,
```

Comment:
Do you need to explicitly match `BoolOp::And` here? Thinking about this, I _think_ you could also match `BoolOp::Or` here and the behavior would be unchanged? A quick comment noting this might be useful because the presence of `And` and not `Or` confused me for a moment. :-)

Reply:
Yeah, I don't _think_ we _need_ to explicitly match `BoolOp::And` here:
- An `or` chained with an `or` (or an `and` chained with an `and`) will just result in a `BoolOp` with three items in the `values` list, since the two operators have the same precedence
- An `and` chained with an `or` (or an `or` chained with an `and`) will both result in an `And` `BoolOp` inside an `Or` `BoolOp`, because the AST reflects the fact that `and` always binds more tightly than `or`.

Why did I leave this the way I did? Two reasons:
1. I got scared I was incorrect in my reasoning ^above, and it seemed "safer" to explicitly check here that it _was_ an `And`, for some definition of the word "safe" :) I could maybe turn it into an assertion instead, to check the invariant holds true -- WDYT?
2. It seemed _slightly_ more readable this way, as now it's obvious to the reader of the code that the subexpression is an `And` `BoolOp`. But that could also be done equally well with an assertion, I suppose!

Reply:
Oh I see now. You don't need to match on the operator at all. Just the existence of a nested `BoolOp` is enough. Right?

> because the AST reflects the fact that and always binds more tightly than or

Ah I didn't realize that.

I think I would go with an `assert!` here. To me, the reasoning is that if your _assumption_ here about the structure of the AST is wrong in some way, then it's likely this code has a bug, right? If so, IMO, it would be better to make that bug loud than to silently continue. i.e., "this tripped a panic" is easier to diagnose as errant behavior than "this lint is producing false negatives." The latter might not be noticed at all.

Reply:
Well, the assertion fails on a bunch of stuff in the ecosystem checks, so I was obviously wrong about something somewhere 🙃

Reply:
Right, of course. It's very possible to have an `Or` inside an `And` -- but it only happens if the `or` is explicitly parenthesized (_otherwise_ the `and` binds more tightly, but the parentheses override that). I.e.:

```py
x = a and (b or c)
```

Reply:
fixed the crash in ab36c8382c4fd663db5f0c1c70b74a25c8fcc81f, and added some more comments :)